File: 백준/Silver/10025. 게으른 백곰/게으른 백곰.java
import java.io.*;
import java.util.*;
public class Main{
    public static void main(String[] args)throws Exception{
        BufferedReader I=new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(I.readLine());
        int n=Integer.parseInt(st.nextToken()),k=Integer.parseInt(st.nextToken()),s=0,lm=(int)(1e6)+1,mxi=0,mni=lm;
        int[] nb=new int[lm];
        k*=2;
        for(int i=0;i<n;i++){
            st=new StringTokenizer(I.readLine());
            int g=Integer.parseInt(st.nextToken()),x=Integer.parseInt(st.nextToken());
            s+=g;
            nb[x]=g;
            if(mxi<x) mxi=x;
            if(mni>x) mni=x;
        }
        n=-1;
        int num=0;
        if(mni+k<lm){
            n=mni+k+1;
            for(int i=mni;i<n;i++){
                num+=nb[i];
            }
            n=num;
        }
        for(int i=mni+k+1;i<=mxi;i++){
            num=num-nb[i-k-1]+nb[i];
            if(num>n) n=num;
        }
        System.out.print(n<0?s:n);
    }
}

File: 백준/Silver/10025. 게으른 백곰/게으른 백곰.py
import sys
def input(): return sys.stdin.readline().rstrip()
n,k=map(int,input().split())
k=k*2
lm=int(1e6)+1
nb=[0]*lm

mxi,mni=0,lm
for i in range(n):
    g,id=map(int,input().split())
    nb[id]=g
    if mxi<id: mxi=id
    if mni>id: mni=id

if mni+k>=lm: n=-1
else:
    num=sum(nb[mni:mni+k+1])
    n=num

for i in range(mni+k+1,mxi+1):
    num=num-nb[i-k-1]+nb[i]
    if num>n: n=num

if n==-1: n=sum(nb[mni:mxi+1])
print(n)